src/librustc_mir/transform/move_up_propagation.rs

```diff
@@ -0,0 +1,596 @@
+                let new_def_stmts: Vec<_> = bbs[def_bb].statements
+                                                        .iter()
+                                                        .enumerate()
+                                                        .map(|(stmt_idx, orig_stmt)| {
+                    if stmt_idx == def_idx {
+                        let new_statement = Statement {
+                            kind: new_statement_kind.clone(),
+                            source_info: orig_stmt.source_info
+                        };
+                        debug!("replacing: {:?} with {:?}.", orig_stmt, new_statement);
+                        new_statement
+                    } else {
+                        orig_stmt.clone()
+                    }
+                }).collect();
+                bbs[def_bb] = BasicBlockData {
+                    statements: new_def_stmts,
+                    terminator: bbs[def_bb].terminator.clone(),
+                    is_cleanup: bbs[def_bb].is_cleanup,
+                };
+
+                // remove DEST = tmp
+                let mut idx_cnt = 0;
+                bbs[use_bb].statements.retain(|orig_stmt| {
+                    let dead = idx_cnt == use_idx;
+                    idx_cnt += 1;
+                    if dead {
+                        debug!("deleting: {:?}", orig_stmt);
+                    }
+                    !dead
+                });
+            }
+
+            // tmp is really a local index (an index into an imaginary vector...)
+            // so we need to convert it back to a Temp
+            if let Some(idx) = mir.from_local_index_to_temp(tmp) {
+                debug!("deleting the decl of: {:?}", idx);
+                dead_temps.insert(idx.index());
+            }
+            if let Some(idx) = mir.from_local_index_to_var(tmp) {
+                debug!("deleting the decl of: {:?}", idx);
+                dead_vars.insert(idx.index());
+            }
+
+            opt_counter += 1;
+        }
+        cleanup_temps(mir, dead_temps);
+        cleanup_vars(mir, dead_vars);
+        debug!("we did {:?} optimizations", opt_counter);
+    }
+}
+
+fn cleanup_temps(mir: &mut Mir, dead_temps: BitVector) {
+    let mut curr = 0;
+    let mut new_vals: IndexVec<Temp, Temp> = IndexVec::with_capacity(mir.temp_decls.len());
+    for i in 0..mir.temp_decls.len() {
+        if dead_temps.contains(i) {
+            // if the value is dead, we don't have any mapping from old to new
+            new_vals.push(Temp::new(0)); // we should never encounter these anyway
+        } else {
+            new_vals.push(Temp::new(curr));
+            curr += 1;
+        }
+    }
+    let mut tr = TempRewriter { new_vals: new_vals };
+    tr.visit_mir(mir);
+    let mut new_decls = IndexVec::new();
+    for (idx, e) in mir.temp_decls.iter_enumerated() {
+        if !dead_temps.contains(idx.index()) {
+            new_decls.push(e.clone());
```

Comment:
Huh, interesting problem. It feels like it'd be nice to leave the "temp indices" in place but just overwrite the entry with some kind of placeholder to indicate it's been removed (we could maybe even maintain a "free list" if we wanted to reuse numbers later). Probably not worth doing in this PR, but maybe worth opening a FIXME. Feels like we should try to make updating the MIR as efficient as possible, and having to do random "renumbering" walks doesn't seem lke it's helping in that respect.